posts/2021/11/pydata-extensibility-vision.md

```diff
@@ -0,0 +1,388 @@
+works with arrays from different libraries. Then it would be possible to
+directly reroute to specific array implementations bypassing compiled code
+paths.
+
+<p align="center">
+    <img
+     alt="A schema using SciPy dispatcher."
+     src="/images/2021/11/scipy-ndimage-dispatcher.png">
+    <i><br>A SciPy module with its own "Dispatcher".</i>
+</p>
+
+<!-- ![scipy.ndimage dispatch schema](https://drive.google.com/uc?export=view&id=1-ZoVxenDufJuUwcONV_mFY0NFNY3wOCq) -->
+
+## Goals, wishes, and constraints
+
+Goal: separate the interface from the implementation and let the dispatching
+system help users and library authors write generic code that works with
+different kinds of arrays.
+
+The new dispatching system implemented for SciPy and scikits should:
+
+* make the API extendable and add the possibility to support alternative array
+  libraries, not just NumPy
+* provide a way to select a different backend for the same array type (examples:
+  `scipy.fft` + alternative CPU FFT libraries like [pyFFTW](https://github.com/pyFFTW/pyFFTW),
+  or scikit-learn + [an Intel optimized
+  implementation](https://github.com/intel/scikit-learn-intelex))
+* be able to extend function signatures in a backward-compatible way on the
+  base library side (all alternative implementations should not break after the
+  change)
+* have a performance overhead that is as low as possible.
+
+When designing API override systems, there are many possible design choices.
+Three major axes of design decisions in the context of NumPy API overrides are
+outlined in [the appendix of NumPy Enhancement Proposal (NEP)
+37](https://numpy.org/neps/nep-0037-array-module.html#appendix-design-choices-for-api-overrides);
+they are also applicable outside of NumPy:
+
+* Opt-in vs. opt-out for users
+* Explicit vs. implicit choice of implementation
+* Local vs. non-local vs. global control.
+
+## A concrete design proposal
+
+In the figure below, we sketch a design for implementing this backend &
+dispatch system in SciPy, scikit-learn, scikit-image and other core projects which
+now rely only on NumPy. It uses CuPy and Dask as examples, with modules that
+exist today. The design is generic though, and will also work for other
+libraries with compatible APIs like PyTorch and JAX once they support this
+system.
+
+We need two dispatching mechanisms: one for the modules with compiled code
+portions, and one for pure Python code.
+The first dispatching mechanism is based on the [`uarray`
+project](https://uarray.org/), which is a backend dispatcher that allows us to
+choose the relevant function implementation at runtime. The second dispatching
+mechanism is based on the ``__array_namespace__`` method from the Array API
+standard, which is a method that allows us to get the Array API implementation
+module specific to CuPy (or any other array library) at runtime.
+``__array_namespace__`` will be available in NumPy 1.22 and CuPy v10.0.
```

Comment:
The second design question to point out is the gray area of dispatching: when something is mostly pure Python code that we want to reuse, _but_ there's some compiled code somewhere. That may be private API, but it needs dispatching still.

Reply:
I thought about it, and we do need to mention it - because it's a fairly obvious gap to maintainers reading this post. I'll have a go at formulating something to address this.

Reply:
Addressing this in a paragraph below the hand-drawn dispatcher diagrams.